File: CS.Algorithm.Training/Task18.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AwesomeAssertions;
using Xunit;

namespace Algorithm.Training;

public class Task18
{
    [Theory]
    [ClassData(typeof(ProductCases))]
    public void Acceptance(
        IEnumerable<IEnumerable<int>> input, 
        Func<int, int, bool> canBeCombined, 
        IEnumerable<IEnumerable<int>> expected)
    {
        input.ProductN(canBeCombined)
            .Should()
            .BeEquivalentTo(expected);
    }
}

public static partial class AdHocExtensions
{
    public static IEnumerable<IEnumerable<T>> ProductN<T>(
        this IEnumerable<IEnumerable<T>> source,
        Func<T, T, bool> canBeCombined)
    {
        return source.Aggregate<IEnumerable<T>, IEnumerable<LinkedList<T>>>(
            [[]],
            (acc, cur) =>
                from prevProductItem in acc
                from item in cur
                where prevProductItem.Last is null || canBeCombined(prevProductItem.Last.Value, item)
                select prevProductItem.With(item));
    }

    private static LinkedList<T> With<T>(this LinkedList<T> list, T item)
    {
        return new LinkedList<T>(list.Append(item));
    }
}

public class ProductCases : TheoryData<IEnumerable<IEnumerable<int>>, Func<int, int, bool>, IEnumerable<IEnumerable<int>>>
{
    public ProductCases()
    {
        Add(
            [
                [1, 2],
                [3, 4]
            ],
            (x, y) => true,
            [
                [1, 3],
                [1, 4],
                [2, 3],
                [2, 4]
            ]);
        
        Add(
            [
                [1, 2, 3],
                [3, 4]
            ],
            (x, y) => (x, y) switch
            {
                (3, _) => y is 3,
                (_, 3) => x is 3,
                _ => true
            },
            [
                [1, 4],
                [2, 4],
                [3, 3]
            ]);
        
        Add(
            [
                [1, 2, 3],
                [3, 4, 5],
                [5, 6, 7]
            ],
            (x, y) => (x, y) switch
            {
                (3, _) => y is 3 or 6 or 7,
                (_, 3) => x is 3,
                (5, _) => y is 5,
                (_, 5) => x is 5 or 1 or 2,
                _ => true
            },
            [
                [1, 4, 6],
                [1, 4, 7],
                [1, 5, 5],
                [2, 4, 6],
                [2, 4, 7],
                [2, 5, 5],
                [3,  3, 6],
                [3, 3, 7]
            ]);
    }
}

File: FS.Algorithm.Training/Task18.fs
module FS.Algorithm.Training.Task18

open Xunit
open Swensen.Unquote

// calque from C# LINQ implementation
let product1 (input: seq<seq<'a>>) =
    (seq { Seq.empty }, input)
    ||> Seq.fold (fun acc cur ->
        seq {
            for prefix in acc do
                for item in cur do
                    yield
                        seq {
                            yield! prefix
                            yield item
                        }
        })

// optimized using lists
let product2 (input: seq<seq<'a>>) =
    (seq { [] }, input)
    ||> Seq.fold (fun acc cur ->
        seq {
            for prefix in acc do
                for item in cur do
                    yield item :: prefix
        })
    |> Seq.map List.rev

// recursive using lists
let rec product3 (input: list<seq<'a>>) =
    match input with
    | [] -> [ [] ]
    | head :: tail ->
        let suffixes = product3 tail
        [ for item in head do
              for rest in suffixes do
                  yield item :: rest ]

// let productWithCondition (condition: 'a -> 'a -> bool) (input: seq<seq<'a>>) : 'a list list =
//     ([[]], input)
//     ||> Seq.fold (fun acc cur ->
//         acc
//         |> List.collect (fun prev ->
//             cur
//             // |> Seq.filter (fun item -> match List.tryLast prev with
//             //                             | Some x -> condition x item
//             //                             | None -> true)
//             |> Seq.map (fun item -> prev @ item))
//         )


type ProductCases() as this =
    inherit TheoryData<int list list, int list list>()
    do this.Add([ [ 1; 2 ]; [ 3; 4 ] ], [ [ 1; 3 ]; [ 1; 4 ]; [ 2; 3 ]; [ 2; 4 ] ])

[<Fact>]
let ``seq equivalence`` () =
    test
        <@
            seq {
                1
                2
                3
            } = seq {
                1
                2
                3
            }
        @>

[<Fact>]
let ``built-in cartesian product only for 2 sequences`` () =
    test <@ ([ 1; 2 ], [ 3; 4 ]) ||> Seq.allPairs |> Seq.toList = [ (1, 3); (1, 4); (2, 3); (2, 4) ] @>

[<Theory>]
[<ClassData(typeof<ProductCases>)>]
let ``get cartesian product`` input expected = test <@ product3  (List.ofSeq input) = expected @>
